--- a/sys/conf/kern.post.mk
+++ b/sys/conf/kern.post.mk
@@ -91 +91 @@
-	    env MKDEP_CPP="${CC} -E" mkdep -a -f .newdep ${ASM_CFLAGS}
+	    /usr/bin/env MKDEP_CPP="${CC} -E" mkdep -a -f .newdep ${ASM_CFLAGS}
@@ -188,2 +188,2 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules obj ; \
-	    env ${MKMODULESENV} ${MAKE} -f Makefile.modules all
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules obj ; \
+	    /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules all
@@ -193,2 +193,2 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules obj ; \
-	    env ${MKMODULESENV} ${MAKE} -f Makefile.modules depend
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules obj ; \
+	    /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules depend
@@ -197 +197 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules clean
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules clean
@@ -200 +200 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules cleandepend
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules cleandepend
@@ -207 +207 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules cleandir
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules cleandir
@@ -210 +210 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules tags
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules tags
@@ -254 +254 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules install
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules install
@@ -258 +258 @@
-	cd $S ; env ${MKMODULESENV} ${MAKE} -f Makefile.modules install
+	cd $S ; /usr/bin/env ${MKMODULESENV} ${MAKE} -f Makefile.modules install
--- a/etc/Makefile
+++ b/etc/Makefile
@@ -383 +383 @@
-	mtree ${MTREE_FSCHG} -deU -f ${.CURDIR}/mtree/BSD.root.dist \
+	mtree ${MTREE_FSCHG} -deU -N ${.CURDIR} -f ${.CURDIR}/mtree/BSD.root.dist \
@@ -385 +385 @@
-	mtree ${MTREE_FSCHG} -deU -f ${.CURDIR}/mtree/BSD.var.dist \
+	mtree ${MTREE_FSCHG} -deU -N ${.CURDIR} -f ${.CURDIR}/mtree/BSD.var.dist \
@@ -387 +387 @@
-	mtree ${MTREE_FSCHG} -deU -f ${.CURDIR}/mtree/BSD.usr.dist \
+	mtree ${MTREE_FSCHG} -deU -N ${.CURDIR} -f ${.CURDIR}/mtree/BSD.usr.dist \
@@ -389 +389 @@
-	mtree ${MTREE_FSCHG} -deU -f ${.CURDIR}/mtree/BSD.include.dist \
+	mtree ${MTREE_FSCHG} -deU -N ${.CURDIR} -f ${.CURDIR}/mtree/BSD.include.dist \
