--- a/src/bin/mountvolume.cpp
+++ b/src/bin/mountvolume.cpp
@@ -379,8 +379,8 @@ MountVolume::RefsReceived(BMessage* message)
 }
 
 
-void
-MountVolume::ArgvReceived(int32 argc, char** argv)
+static void
+process_arguments(int32 argc, char** argv)
 {
 	MountVisitor mountVisitor;
 	PrintPartitionsVisitor printPartitionsVisitor;
@@ -558,6 +558,13 @@ MountVolume::ArgvReceived(int32 argc, char** argv)
 }
 
 
+void
+MountVolume::ArgvReceived(int32 argc, char** argv)
+{
+	process_arguments(argc, argv);
+}
+
+
 void
 MountVolume::ReadyToRun()
 {
@@ -575,8 +582,12 @@ MountVolume::ReadyToRun()
 
 
 int
-main()
+main(int argc, char** argv)
 {
+	// Command line operation does not need a connection to app_server.
+	if (argc > 1)
+		process_arguments(argc, argv);
+
 	MountVolume mountVolume;
 	mountVolume.Run();
 	return 0;
