diff --git a/tools/cctools/ld64/src/3rd/include/CommonCrypto/CommonDigest.h b/tools/cctools/ld64/src/3rd/include/CommonCrypto/CommonDigest.h
index 1f5a9e5..b45b817 100644
--- a/tools/cctools/ld64/src/3rd/include/CommonCrypto/CommonDigest.h
+++ b/tools/cctools/ld64/src/3rd/include/CommonCrypto/CommonDigest.h
@@ -1,3 +1,6 @@
+#ifndef LD64_COMMON_CRYPTO_COMMON_DIGEST_H
+#define LD64_COMMON_CRYPTO_COMMON_DIGEST_H
+
 #ifdef __APPLE__
 
 #include_next <CommonCrypto/CommonDigest.h>
@@ -46,3 +49,5 @@ static unsigned char *CC_MD5(const void *data, unsigned long nbytes,
 }
 
 #endif /* __APPLE__ */
+
+#endif /* LD64_COMMON_CRYPTO_COMMON_DIGEST_H */
diff --git a/tools/cctools/ld64/src/ld/parsers/CMakeLists.txt b/tools/cctools/ld64/src/ld/parsers/CMakeLists.txt
index 12ac0df..4f7beb7 100644
--- a/tools/cctools/ld64/src/ld/parsers/CMakeLists.txt
+++ b/tools/cctools/ld64/src/ld/parsers/CMakeLists.txt
@@ -10,6 +10,7 @@ target_sources(host_ld_libparsers PRIVATE
 )
 
 target_compile_definitions(host_ld_libparsers PRIVATE __DARWIN_UNIX03)
+target_compile_options(host_ld_libparsers PRIVATE -fblocks)
 target_include_directories(host_ld_libparsers PRIVATE
     ../../../../include
     ../../../../include/foreign
@@ -20,5 +21,5 @@ target_include_directories(host_ld_libparsers PRIVATE
     ..
 )
 target_link_libraries(host_ld_libparsers PRIVATE host_ld_helper)
-set_property(TARGET host_ld_libparsers PROPERTY CXX_STANDARD 11)
+set_property(TARGET host_ld_libparsers PROPERTY CXX_STANDARD 17)
 suppress_all_warnings(host_ld_libparsers)
diff --git a/tools/cctools/ld64/src/ld/parsers/generic_dylib_file.hpp b/tools/cctools/ld64/src/ld/parsers/generic_dylib_file.hpp
index b1ab029..b07f474 100644
--- a/tools/cctools/ld64/src/ld/parsers/generic_dylib_file.hpp
+++ b/tools/cctools/ld64/src/ld/parsers/generic_dylib_file.hpp
@@ -30,6 +30,7 @@
 #include "Options.h"
 #include <unordered_map>
 #include <unordered_set>
+#include <memory>
 
 namespace generic {
 namespace dylib {
diff --git a/tools/cctools/ld64/src/ld/passes/CMakeLists.txt b/tools/cctools/ld64/src/ld/passes/CMakeLists.txt
index 3010c4a..740c34d 100644
--- a/tools/cctools/ld64/src/ld/passes/CMakeLists.txt
+++ b/tools/cctools/ld64/src/ld/passes/CMakeLists.txt
@@ -19,6 +19,7 @@ target_sources(host_ld_libpasses PRIVATE
 )
 
 target_compile_definitions(host_ld_libpasses PRIVATE __DARWIN_UNIX03)
+target_compile_options(host_ld_libpasses PRIVATE -fblocks)
 target_include_directories(host_ld_libpasses PRIVATE
     ../../../../include
     ../../../../include/foreign
@@ -29,5 +30,5 @@ target_include_directories(host_ld_libpasses PRIVATE
     ../../3rd
     ../../..
 )
-set_property(TARGET host_ld_libpasses PROPERTY CXX_STANDARD 11)
+set_property(TARGET host_ld_libpasses PROPERTY CXX_STANDARD 17)
 suppress_all_warnings(host_ld_libpasses)
