diff --git a/tools/dtrace_ctf/tools/ctf.c b/tools/dtrace_ctf/tools/ctf.c
index a3fafc4..294756c 100644
--- a/tools/dtrace_ctf/tools/ctf.c
+++ b/tools/dtrace_ctf/tools/ctf.c
@@ -1241,6 +1241,6 @@ decompress_ctf(caddr_t cbuf, size_t cbufsz, caddr_t dbuf, size_t dbufsz)
 	    (rc = inflateEnd(&zstr)) != Z_OK) {
 		warning("CTF decompress zlib error %s\n", zError(rc));
-		return (NULL);
+		return (0);
 	}
 
 	debug(3, "reflated %lu bytes to %lu, pointer at %d\n",
