diff --git a/tools/dtrace_ctf/tools/compare.c b/tools/dtrace_ctf/tools/compare.c
index 3894343..3b2e911 100644
--- a/tools/dtrace_ctf/tools/compare.c
+++ b/tools/dtrace_ctf/tools/compare.c
@@ -39,7 +39,7 @@
 
 #include "ctftools.h"
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 extern const
 #endif /* __APPLE__ */ 
 char *progname;
diff --git a/tools/dtrace_ctf/tools/ctfconvert.c b/tools/dtrace_ctf/tools/ctfconvert.c
index 562e406..4f457f6 100644
--- a/tools/dtrace_ctf/tools/ctfconvert.c
+++ b/tools/dtrace_ctf/tools/ctfconvert.c
@@ -105,7 +105,7 @@ file_read(const char *filename, const char *cumatch, int verbose, int ignore_non
 		    elf_errmsg(-1));
 	}
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	source_types = built_source_types(elf, filename);
 #else
 	source_types = SOURCE_C | SOURCE_S;
@@ -137,7 +137,7 @@ file_read(const char *filename, const char *cumatch, int verbose, int ignore_non
 			    filename);
 			exit(0);
 		}
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		/* produce an empty output */
 		*td = tdata_new();
 		rc = 1;
diff --git a/tools/dtrace_ctf/tools/ctfmerge.c b/tools/dtrace_ctf/tools/ctfmerge.c
index af7fff2..857e9fb 100644
--- a/tools/dtrace_ctf/tools/ctfmerge.c
+++ b/tools/dtrace_ctf/tools/ctfmerge.c
@@ -23,7 +23,7 @@
  * Use is subject to license terms.
  */
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #include <stdio.h>
 #include <stdlib.h>
 #include <unistd.h>
@@ -87,7 +87,7 @@ usage(void)
 	    "       %s [-fgstv] -l label | -L labelenv -o outfile -w withfile "
 	    "file ...\n"
 	    "       %s [-g] -c srcfile destfile\n"
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	    "       %s [-fgstv] -l label | -L labelenv -o master_macho_file -Z raw_ctf_outfile file ...\n"
 #endif
 	    "\n"
@@ -153,7 +153,7 @@ main(int argc, char **argv)
 	tdata_t *mstrtd, *savetd;
 	char *uniqfile = NULL, *uniqlabel = NULL;
 	char *withfile = NULL;
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	char *raw_ctf_file = NULL;
 #endif
 	char *label = NULL;
@@ -171,7 +171,7 @@ main(int argc, char **argv)
 		debug_level = atoi(getenv("CTFMERGE_DEBUG_LEVEL"));
 
 	err = 0;
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	while ((c = getopt(argc, argv, ":cd:D:fgl:L:o:tvw:sZ:")) != EOF) {
 #else
 	while ((c = getopt(argc, argv, ":cd:D:fgl:L:o:tvw:s")) != EOF) {
@@ -223,7 +223,7 @@ main(int argc, char **argv)
 			/* use the dynsym rather than the symtab */
 			dynsym = CTF_USE_DYNSYM;
 			break;
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		case 'Z':
 			/* Write raw CTF data by itself */
 			raw_ctf_file = optarg;
@@ -257,7 +257,7 @@ main(int argc, char **argv)
 			err++;
 	}
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	if ((uniqfile != NULL || withfile != NULL) && raw_ctf_file != NULL)
 		err++;
 #endif
@@ -283,7 +283,7 @@ main(int argc, char **argv)
 	if (outfile && access(outfile, R_OK|W_OK) != 0)
 		terminate("Cannot open output file %s for r/w", outfile);
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	if (raw_ctf_file && access(raw_ctf_file, F_OK) != -1)
 		terminate("Raw CTF output file %s already exists", raw_ctf_file);
 #endif
@@ -422,7 +422,7 @@ main(int argc, char **argv)
 		savetd = mstrtd;
 	}
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	tmpname = mktmpname(outfile, ".ctf");
 	write_ctf(savetd, outfile, tmpname,
 		  CTF_COMPRESS | write_fuzzy_match | dynsym | keep_stabs);
diff --git a/tools/dtrace_ctf/tools/ctftools.h b/tools/dtrace_ctf/tools/ctftools.h
index 8c832c9..d8e8881 100644
--- a/tools/dtrace_ctf/tools/ctftools.h
+++ b/tools/dtrace_ctf/tools/ctftools.h
@@ -395,6 +395,6 @@ tdata_t *ctfmerge_done(void);
 #define	CTF_KEEP_STABS	0x08 /* keep .stabs sections */
 #define	CTF_MINIMIZE	0x10 /* create minimal symbols & CTF output file */
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 #define CTF_RAW_OUTPUT	(1<<30) /* output raw CTF data only, no segment/section data, and do not copy original file */
 #endif /* __APPLE__ */
 
diff --git a/tools/dtrace_ctf/tools/dump.c b/tools/dtrace_ctf/tools/dump.c
index 6d88088..0a8bada 100644
--- a/tools/dtrace_ctf/tools/dump.c
+++ b/tools/dtrace_ctf/tools/dump.c
@@ -24,7 +24,7 @@
  * Use is subject to license terms.
  */
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #include <sys/types.h>
 #include <sys/sysmacros.h>
 #include <sys/stat.h>
@@ -44,6 +44,7 @@ int debug_level;
 const char *progname = "ctfdump";
 #endif /* __APPLE__ */
 
+#include <string.h>
 #include <strings.h>
 #include <unistd.h>
 #include <stdlib.h>
@@ -58,7 +59,7 @@ const char *progname = "ctfdump";
 #include "utils.h"
 #include "symbol.h"
 
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 extern caddr_t write_buffer(ctf_header_t *h, ctf_buf_t *buf, size_t *resszp);
 extern caddr_t write_compressed_buffer(ctf_header_t *h, ctf_buf_t *buf, size_t *resszp);
 struct ctf_buf {
@@ -262,7 +263,7 @@ print_labeltable(const ctf_header_t *hp, const ctf_data_t *cd)
 	return (E_SUCCESS);
 }
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 static GElf_Sym *
 gelf_getsym_macho(Elf_Data * data, int ndx, GElf_Sym * sym, const char *base)
 {
@@ -377,7 +378,7 @@ next_sym(const ctf_data_t *cd, const int symidx, const uchar_t matchtype,
 		char *name;
 		int type;
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		if (cd->sh_link == SHN_MACHO) { /* Underlying file is Mach-o */
 			if (gelf_getsym_macho(cd->cd_symdata, i, &sym, (const char *)(cd->cd_strdata->d_buf)) == 0)
 				return (-1);
@@ -1021,7 +1022,7 @@ print_stats(void)
 static int
 print_usage(FILE *fp, int verbose)
 {
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	(void) fprintf(fp, "Usage: %s [-dfhlsSt] [-u file] file\n", getpname());
 #else
 	(void) fprintf(fp, "Usage: %s [-dfhlrsSt] [-u file] file\n", progname);
@@ -1033,7 +1034,7 @@ print_usage(FILE *fp, int verbose)
 		    "\t-f  dump function section\n"
 		    "\t-h  dump file header\n"
 		    "\t-l  dump label table\n"
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		    "\t-r  remove listed symbols from CTF data\n" 
 #endif /* __APPLE__ */
 		    "\t-s  dump string table\n"
@@ -1156,6 +1157,6 @@ skiploop:
 		 * is zero or if there's an error reading the section header.
 		 */
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 		if (gelf_getshdr(ctfscn, &ctfshdr) != NULL &&
 		    ctfshdr.sh_link != 0) {
 			symscn = elf_getscn(elf, ctfshdr.sh_link);
@@ -1178,7 +1179,7 @@ skiploop:
 			Elf_Scn *symstrscn;
 
 			if (gelf_getshdr(symscn, &shdr) != NULL) {
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 				symstrscn = elf_getscn(elf, shdr.sh_link);
 
 				cd.cd_nsyms = shdr.sh_size / shdr.sh_entsize;
@@ -1297,7 +1298,7 @@ skiploop:
 		cd.cd_ctflen = hp->cth_stroff + hp->cth_strlen;
 	}
 
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	if (doclean) {
 		int nsyms = argc - 1 - optind;
 		char **syms = &argv[optind];
diff --git a/tools/dtrace_ctf/tools/hash.c b/tools/dtrace_ctf/tools/hash.c
index 635a5e1..85de599 100644
--- a/tools/dtrace_ctf/tools/hash.c
+++ b/tools/dtrace_ctf/tools/hash.c
@@ -28,7 +28,7 @@
  * Routines for manipulating hash tables
  */
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #include <stdio.h>
 #include <stdlib.h>
 #include <strings.h>
diff --git a/tools/dtrace_ctf/tools/input.c b/tools/dtrace_ctf/tools/input.c
index e78a359..9aa0f98 100644
--- a/tools/dtrace_ctf/tools/input.c
+++ b/tools/dtrace_ctf/tools/input.c
@@ -32,6 +32,7 @@
 #include <fcntl.h>
 #include <unistd.h>
 #include <gelf.h>
+#include <string.h>
 #include <strings.h>
 #include <sys/types.h>
 
@@ -105,7 +106,7 @@ read_file(Elf *elf, char *file, char *label, read_cb_f *func, void *arg,
 
 	/* Reconstruction of type tree */
 	if ((si = symit_new(elf, file)) == NULL) {
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 		warning("%s has no symbol table - skipping", file);
 #endif
 		return (0);
@@ -144,7 +145,7 @@ read_ctf_common(char *file, char *label, read_cb_f *func, void *arg,
 
 	switch (elf_kind(elf)) {
 	case ELF_K_ELF:
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	case ELF_K_MACHO: /* Underlying file is Mach-o */
 #endif /* __APPLE__ */
 		found = read_file(elf, file, label,
@@ -217,7 +218,7 @@ count_files(char **files, int n)
 
 		switch (elf_kind(elf)) {
 		case ELF_K_ELF:
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		case ELF_K_MACHO: /* Underlying file is Mach-o */
 #endif /* __APPLE__ */
 			nfiles++;
@@ -267,7 +268,7 @@ symit_new(Elf *elf, const char *file)
 	    (si->si_symd = elf_getdata(scn, NULL)) == NULL)
 		elfterminate(file, "Cannot read .symtab");
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	if ((scn = elf_getscn(elf, si->si_shdr.sh_link)) == NULL ||
 	    (si->si_strd = elf_getdata(scn, NULL)) == NULL)
 		elfterminate(file, "Cannot read strings for .symtab");
@@ -309,7 +310,7 @@ symit_curfile(symit_data_t *si)
 	return (si->si_curfile);
 }
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 #include <mach-o/loader.h>
 #include <mach-o/nlist.h>
 
@@ -416,7 +417,7 @@ symit_next(symit_data_t *si, int type)
 	int check_sym = (type == STT_OBJECT || type == STT_FUNC);
 
 	for (; si->si_next < si->si_nument; si->si_next++) {
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 		gelf_getsym(si->si_symd, si->si_next, &si->si_cursym);
 		gelf_getsym(si->si_symd, si->si_next, &sym);
 #else
diff --git a/tools/dtrace_ctf/tools/merge.c b/tools/dtrace_ctf/tools/merge.c
index 604598b..c05b17d 100644
--- a/tools/dtrace_ctf/tools/merge.c
+++ b/tools/dtrace_ctf/tools/merge.c
@@ -124,7 +124,7 @@
 #include "memory.h"
 #include "traverse.h"
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 #include <unistd.h>
 #include <signal.h>
 #endif /* __APPLE__ */
@@ -1320,7 +1320,7 @@ tdesc_ops_t tdesc_ops[] = {
  * this comment will not live to see the stock hit five.
  */
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #pragma init(bigheap)
 static size_t maxpgsize = 0x400000;
 #endif /* __APPLE__ */
@@ -1329,7 +1329,7 @@ static size_t maxpgsize = 0x400000;
 #define	MERGE_PHASE1_MAX_SLOTS		5
 #define	MERGE_INPUT_THROTTLE_LEN	10
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 static void
 bigheap(void)
 {
@@ -1735,7 +1735,7 @@ wq_init(workqueue_t *wq, int nfiles)
 
 	for (i = 0; i < nslots; i++) {
 		pthread_mutex_init(&wq->wq_wip[i].wip_lock, NULL);
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 		pthread_cond_init(&wq->wq_wip[i].wip_cv, NULL); /* Omitted on Solaris!?! */
 #endif /* __APPLE__ */
 		wq->wq_wip[i].wip_batchid = wq->wq_next_batchid++;
diff --git a/tools/dtrace_ctf/tools/output.c b/tools/dtrace_ctf/tools/output.c
index d667e8b..b05dba1 100644
--- a/tools/dtrace_ctf/tools/output.c
+++ b/tools/dtrace_ctf/tools/output.c
@@ -30,6 +30,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <strings.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -44,7 +45,7 @@
 #include "traverse.h"
 #include "symbol.h"
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 #include <mach-o/loader.h>
 #include <mach-o/nlist.h>
 #include <sys/mman.h>
@@ -439,7 +440,7 @@ sort_iidescs(Elf *elf, const char *file, tdata_t *td, int fuzzymatch,
 
 	if ((stidx = findelfsecidx(elf, file,
 	    dynsym ? ".dynsym" : ".symtab")) < 0)
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 		terminate("%s: Can't open symbol table\n", file);
 #else
 	return (iiburst_new(td, 0)); /* missing symbol table is most likely an empty binary,
@@ -450,7 +451,7 @@ sort_iidescs(Elf *elf, const char *file, tdata_t *td, int fuzzymatch,
 	gelf_getshdr(scn, &shdr);
 	nent = shdr.sh_size / shdr.sh_entsize;
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	scn = elf_getscn(elf, shdr.sh_link);
 	strdata = elf_getdata(scn, NULL);
 #else
@@ -470,7 +471,7 @@ sort_iidescs(Elf *elf, const char *file, tdata_t *td, int fuzzymatch,
 
 	iiburst = iiburst_new(td, nent);
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	for (i = 0; i < nent; i++) {
 		GElf_Sym sym;
 		iidesc_t **tolist;
@@ -1088,7 +1089,7 @@ write_ctf(tdata_t *td, const char *curname, const char *newname, int flags)
 	if ((telf = elf_begin(tfd, ELF_C_WRITE, NULL)) == NULL)
 		elfterminate(curname, "Cannot write");
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || defined(PUREDARWIN_HOST)
 	if (ELFCLASS32 == elf->ed_class) {
 		data = make_ctf_data(td, elf, curname, &len, flags);
 		if (flags & CTF_RAW_OUTPUT) {
diff --git a/tools/dtrace_ctf/tools/strtab.cpp b/tools/dtrace_ctf/tools/strtab.cpp
index 42aca0a..e3e3e5f 100644
--- a/tools/dtrace_ctf/tools/strtab.cpp
+++ b/tools/dtrace_ctf/tools/strtab.cpp
@@ -24,7 +24,7 @@
  * All rights reserved.
  */
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #include <sys/types.h>
 #include <sys/sysmacros.h>
 #include <strings.h>
diff --git a/tools/dtrace_ctf/tools/tdata.c b/tools/dtrace_ctf/tools/tdata.c
index 20ae26b..2667f94 100644
--- a/tools/dtrace_ctf/tools/tdata.c
+++ b/tools/dtrace_ctf/tools/tdata.c
@@ -29,6 +29,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <strings.h>
 #include <pthread.h>
 
diff --git a/tools/dtrace_ctf/tools/util.c b/tools/dtrace_ctf/tools/util.c
index 751d52b..635a789 100644
--- a/tools/dtrace_ctf/tools/util.c
+++ b/tools/dtrace_ctf/tools/util.c
@@ -138,7 +138,7 @@ terminate(const char *format, ...)
 	va_list ap;
 
 	va_start(ap, format);
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	whine("ERROR", format, ap);
 #else
     /*
@@ -168,7 +168,7 @@ aborterr(char *format, ...)
 	whine("ERROR", format, ap);
 	va_end(ap);
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 	abort();
 #else
 	/*
diff --git a/tools/dtrace_ctf/tools/utils.c b/tools/dtrace_ctf/tools/utils.c
index e471161..083d15d 100644
--- a/tools/dtrace_ctf/tools/utils.c
+++ b/tools/dtrace_ctf/tools/utils.c
@@ -36,7 +36,7 @@
 
 static const char *pname;
 
-#if !defined(__APPLE__)
+#if !defined(__APPLE__) && !defined(PUREDARWIN_HOST)
 #pragma init(getpname)
 const char *
 getpname(void)
