diff --git a/src/Kernel/xnu/CMakeLists.txt b/src/Kernel/xnu/CMakeLists.txt
index f12b7a0..243e048 100644
--- a/src/Kernel/xnu/CMakeLists.txt
+++ b/src/Kernel/xnu/CMakeLists.txt
@@ -1,5 +1,17 @@
 find_program(dsymutil_path NAMES dsymutil llvm-dsymutil REQUIRED)
 
+set(PUREDARWIN_BUILD_JOBS 1 CACHE STRING "Parallel XNU build jobs")
+set(xnu_make_environment
+    SYSCTL_HW_PHYSICALCPU=${PUREDARWIN_BUILD_JOBS}
+    SYSCTL_HW_LOGICALCPU=${PUREDARWIN_BUILD_JOBS}
+    MAKEJOBS=--jobs=${PUREDARWIN_BUILD_JOBS}
+    MEMORY_SIZE=268435456
+    SDKROOT_RESOLVED=/
+    HOST_SDKROOT_RESOLVED=/
+    PLATFORM=MacOSX
+    HOST_OS_VERSION=11.0
+)
+
 file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/xnu_header_install)
 externalproject_add(xnu_headers.extproj
     PREFIX ${CMAKE_CURRENT_BINARY_DIR}/xnu_header_build
@@ -25,7 +37,8 @@ externalproject_add(xnu_headers.extproj
             -D DSYMUTIL_PATH=${dsymutil_path}
             -P ${CMAKE_CURRENT_SOURCE_DIR}/preprocess_files.cmake
     COMMAND
-        make -C <SOURCE_DIR> installhdrs ARCH_CONFIGS=X86_64 BUILD_LTO=0
+        make -C <SOURCE_DIR> installhdrs ${xnu_make_environment}
+            ARCH_CONFIGS=X86_64 BUILD_LTO=0
             BUILD_WERROR=0 DO_CTFMERGE=0 KERNEL_CONFIGS=RELEASE
             SRCROOT=${CMAKE_CURRENT_SOURCE_DIR} OBJROOT=<BINARY_DIR>
             DSTROOT=${CMAKE_CURRENT_BINARY_DIR}/xnu_header_install
@@ -92,7 +105,8 @@ externalproject_add(xnu
             -D PTHREAD_HEADER_PATH=$<TARGET_PROPERTY:pthread_common_headers,INTERFACE_INCLUDE_DIRECTORIES>
             -P ${CMAKE_CURRENT_SOURCE_DIR}/preprocess_files.cmake
     COMMAND
-        make -C <SOURCE_DIR> install ARCH_CONFIGS=X86_64 BUILD_LTO=0
+        make -C <SOURCE_DIR> install ${xnu_make_environment}
+            ARCH_CONFIGS=X86_64 BUILD_LTO=0
             BUILD_WERROR=0 DO_CTFMERGE=0 KERNEL_CONFIGS=RELEASE
             SRCROOT=${CMAKE_CURRENT_SOURCE_DIR} OBJROOT=<BINARY_DIR>
             SYMROOT=${CMAKE_CURRENT_BINARY_DIR}/xnu_sym DSTROOT=${CMAKE_CURRENT_BINARY_DIR}/xnu
