diff --git a/src/Kernel/xnu/CMakeLists.txt b/src/Kernel/xnu/CMakeLists.txt
index 21306b0..cab7e7e 100644
--- a/src/Kernel/xnu/CMakeLists.txt
+++ b/src/Kernel/xnu/CMakeLists.txt
@@ -12,5 +12,7 @@ set(xnu_make_environment
     HOST_OS_VERSION=11.0
     PUREDARWIN_HOST_CCTOOLS_INCLUDE=${PUREDARWIN_SOURCE_DIR}/tools/cctools/include
+    PUREDARWIN_HOST_LIBMACHO=$<TARGET_FILE:host_libmacho>
+    PUREDARWIN_HOST_LIBSTUFF=$<TARGET_FILE:host_libstuff>
 )
 
 file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/xnu_header_install)
diff --git a/src/Kernel/xnu/SETUP/kextsymboltool/Makefile b/src/Kernel/xnu/SETUP/kextsymboltool/Makefile
index b95e4dc..ca36fc0 100644
--- a/src/Kernel/xnu/SETUP/kextsymboltool/Makefile
+++ b/src/Kernel/xnu/SETUP/kextsymboltool/Makefile
@@ -17,9 +17,12 @@ WARNFLAGS = -Wall
 
 LDFLAGS = -isysroot $(HOST_SDKROOT) -mmacosx-version-min=$(HOST_OS_VERSION) -lstdc++
+ifneq ($(PUREDARWIN_HOST_LIBMACHO),)
+LDLIBS = $(PUREDARWIN_HOST_LIBMACHO) $(PUREDARWIN_HOST_LIBSTUFF) -luuid
+endif
 
 kextsymboltool: $(OBJS)
 	@$(LOG_HOST_LD) "$@"
-	$(_v)$(HOST_CC) $(LDFLAGS) -o $@ $^
+	$(_v)$(HOST_CC) $(LDFLAGS) -o $@ $^ $(LDLIBS)
 
 %.o: %.c
 	@$(LOG_HOST_CC) "$@"
