diff --git a/src/Kernel/xnu/makedefs/MakeInc.top b/src/Kernel/xnu/makedefs/MakeInc.top
index c5a9899..9cf3325 100644
--- a/src/Kernel/xnu/makedefs/MakeInc.top
+++ b/src/Kernel/xnu/makedefs/MakeInc.top
@@ -399,13 +399,15 @@ $(eval $(generated_top_level_build_setup))
 # invalidate current kernel in $(SYMROOT). Timestamp must be +1 from a previous kernel build
 setup: build_setup_bootstrap
 	$(_v)$(TOUCH) $(OBJROOT)/.mach_kernel.timestamp.new
-	$(_v)while [ \! $(OBJROOT)/.mach_kernel.timestamp.new -nt $(OBJROOT)/.mach_kernel.timestamp ]; do \
+	$(_v)while [ -e $(OBJROOT)/.mach_kernel.timestamp ] && \
+	    [ \! $(OBJROOT)/.mach_kernel.timestamp.new -nt $(OBJROOT)/.mach_kernel.timestamp ]; do \
 		$(SLEEP) 1;					\
 		$(TOUCH) $(OBJROOT)/.mach_kernel.timestamp.new;	\
 	done
 	$(_v)$(MV) $(OBJROOT)/.mach_kernel.timestamp.new $(OBJROOT)/.mach_kernel.timestamp
 	$(_v)$(TOUCH) $(OBJROOT)/.symbolset.timestamp.new
-	$(_v)while [ \! $(OBJROOT)/.symbolset.timestamp.new -nt $(OBJROOT)/.symbolset.timestamp ]; do \
+	$(_v)while [ -e $(OBJROOT)/.symbolset.timestamp ] && \
+	    [ \! $(OBJROOT)/.symbolset.timestamp.new -nt $(OBJROOT)/.symbolset.timestamp ]; do \
 		$(SLEEP) 1;					\
 		$(TOUCH) $(OBJROOT)/.symbolset.timestamp.new;	\
 	done
