diff --git a/src/Kernel/xnu/SETUP/config/doconf b/src/Kernel/xnu/SETUP/config/doconf
--- a/src/Kernel/xnu/SETUP/config/doconf
+++ b/src/Kernel/xnu/SETUP/config/doconf
@@ -188,7 +188,7 @@ foreach SYS ($argv)
     cat $MASTER $MASTER_CPU - \
         $MASTER $MASTER_CPU \
     | \
-    unifdef -t -DPLATFORM_${PLATFORM} -DCPU_$cpu -DSOC_CONFIG_${SOC_CONFIG} -DSYS_${SYS} - \
+    $UNIFDEF -t -DPLATFORM_${PLATFORM} -DCPU_$cpu -DSOC_CONFIG_${SOC_CONFIG} -DSYS_${SYS} - \
     | \
     sed -n \
 	-e "/^+/{" \
