diff --git a/src/Kernel/xnu/osfmk/ipc/mach_kernelrpc.c b/src/Kernel/xnu/osfmk/ipc/mach_kernelrpc.c
--- a/src/Kernel/xnu/osfmk/ipc/mach_kernelrpc.c
+++ b/src/Kernel/xnu/osfmk/ipc/mach_kernelrpc.c
@@ -319,7 +319,7 @@ _kernelrpc_mach_port_get_attributes_trap(struct _kernelrpc_mach_port_get_attribut
 #define MACH_PORT_INFO_STACK_LIMIT 80 // current size is 68 == 17 * sizeof(integer_t)
 	_Static_assert(sizeof(MACH_PORT_INFO_OUT) < MACH_PORT_INFO_STACK_LIMIT,
 	    "mach_port_info_t has grown significantly, reevaluate stack usage");
-	const mach_msg_type_number_t max_count = (sizeof(MACH_PORT_INFO_OUT) / sizeof(MACH_PORT_INFO_OUT[0]));
+	enum { max_count = (sizeof(MACH_PORT_INFO_OUT) / sizeof(MACH_PORT_INFO_OUT[0])) };
 	typeof(MACH_PORT_INFO_OUT[0]) info[max_count];
 
 	/*
