diff --git a/src/Kernel/CMakeLists.txt b/src/Kernel/CMakeLists.txt
--- a/src/Kernel/CMakeLists.txt
+++ b/src/Kernel/CMakeLists.txt
@@ -1,4 +1,30 @@
 add_subdirectory(Extensions)
 add_subdirectory(libkmod)
 add_subdirectory(libfirehose_kernel)
+
+if(PUREDARWIN_COMPILER_RT_SOURCE_DIR)
+    add_darwin_static_library(compiler_rt_profile)
+    target_sources(compiler_rt_profile PRIVATE
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfiling.c
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfilingBuffer.c
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfilingPlatformDarwin.c
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfilingWriter.c
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfilingInternal.c
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile/InstrProfilingVersionVar.c
+    )
+    target_include_directories(compiler_rt_profile PRIVATE
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/include
+        ${PUREDARWIN_COMPILER_RT_SOURCE_DIR}/lib/profile
+        ${PUREDARWIN_SOURCE_DIR}/src/Libraries/libSystem/libsystem_kernel/include
+    )
+    target_compile_definitions(compiler_rt_profile PRIVATE VISIBILITY_HIDDEN)
+    target_compile_options(compiler_rt_profile PRIVATE
+        -ibuiltininc -fPIC -O3 -fvisibility=hidden -fomit-frame-pointer
+    )
+    target_link_libraries(compiler_rt_profile PRIVATE
+        AvailabilityHeaders libc_public_headers pthread_common_headers
+        xnu_headers xnu_kernel_headers
+    )
+endif()
+
 add_subdirectory(xnu)
diff --git a/src/Kernel/xnu/CMakeLists.txt b/src/Kernel/xnu/CMakeLists.txt
--- a/src/Kernel/xnu/CMakeLists.txt
+++ b/src/Kernel/xnu/CMakeLists.txt
@@ -107,6 +107,7 @@ externalproject_add(xnu
             -D DSYMUTIL_PATH=${dsymutil_path}
             -D FIREHOSE_KERNEL_LIBRARY_PATH=$<TARGET_FILE:libfirehose_kernel>
             -D FIREHOSE_KERNEL_HEADER_PATH=$<TARGET_PROPERTY:libfirehose_kernel,INTERFACE_INCLUDE_DIRECTORIES>
+            -D PROFILE_RUNTIME_LIBRARY_PATH=$<TARGET_FILE:compiler_rt_profile>
             -D PTHREAD_HEADER_PATH=$<TARGET_PROPERTY:pthread_common_headers,INTERFACE_INCLUDE_DIRECTORIES>
             -P ${CMAKE_CURRENT_SOURCE_DIR}/preprocess_files.cmake
     COMMAND
@@ -121,7 +122,7 @@ externalproject_add(xnu
     USES_TERMINAL_BUILD TRUE
 )
 add_dependencies(xnu
-    xnu_headers libfirehose_kernel host_codesign_allocate host_ctfconvert host_ctfmerge
+    xnu_headers libfirehose_kernel compiler_rt_profile host_codesign_allocate host_ctfconvert host_ctfmerge
     migcom host_strip host_lipo host_nm unifdef host_nmedit availability.pl host_ld
     dyld_availability_headers
 )
diff --git a/src/Kernel/xnu/cmake/MakeInc.def.in b/src/Kernel/xnu/cmake/MakeInc.def.in
--- a/src/Kernel/xnu/cmake/MakeInc.def.in
+++ b/src/Kernel/xnu/cmake/MakeInc.def.in
@@ -754,7 +754,7 @@ LDFLAGS_KERNEL_ONLY   =   \
 # Default runtime libraries to be linked with the kernel
 #
 LD_KERNEL_LIBS	   = -lcc_kext
-LD_KERNEL_ARCHIVES = $(LDFLAGS_KERNEL_SDK) -lfirehose_kernel
+LD_KERNEL_ARCHIVES = $(LDFLAGS_KERNEL_SDK) -lfirehose_kernel @PROFILE_RUNTIME_LIBRARY_PATH@
 
 #
 # DTrace support
