diff --git a/src/Kernel/xnu/libkern/c++/OSKext.cpp b/src/Kernel/xnu/libkern/c++/OSKext.cpp
index a9dbc82e..f67c2339 100644
--- a/src/Kernel/xnu/libkern/c++/OSKext.cpp
+++ b/src/Kernel/xnu/libkern/c++/OSKext.cpp
@@ -6259,7 +6259,8 @@ OSKext::loadExecutable(bool startFlag, OSKextDelayAutounloadFlag delayAutounload
 	}
 
 	/* <rdar://problem/21444003> all callers must be entitled */
-	if (FALSE == IOTaskHasEntitlement(current_task(), kOSKextCollectionManagementEntitlement)) {
+	if (current_task() != kernel_task &&
+	    FALSE == IOTaskHasEntitlement(current_task(), kOSKextCollectionManagementEntitlement)) {
 		OSKextLog(this,
 		    kOSKextLogErrorLevel | kOSKextLogLoadFlag,
 		    "Not entitled to link kext '%s'",
