diff --git a/src/Kernel/xnu/config/MASTER.x86_64 b/src/Kernel/xnu/config/MASTER.x86_64
index 0c58f2df..2ebd770c 100644
--- a/src/Kernel/xnu/config/MASTER.x86_64
+++ b/src/Kernel/xnu/config/MASTER.x86_64
@@ -40,7 +40,7 @@
 #  IOKIT_RELEASE =  [ IOKIT_BASE ]
 #  IOKIT_DEV =      [ IOKIT_BASE iotracking ]
 #  IOKIT_DEBUG =    [ IOKIT_BASE iotracking ]
-#  LIBKERN_BASE =   [ libkerncpp config_blocks config_kec_fips crypto_sha2 config_img4 ]
+#  LIBKERN_BASE =   [ libkerncpp config_blocks config_kxld config_kec_fips crypto_sha2 config_img4 ]
 #  LIBKERN_RELEASE =[ LIBKERN_BASE zlib ]
 #  LIBKERN_DEV =    [ LIBKERN_BASE zlib iotracking ]
 #  LIBKERN_DEBUG =  [ LIBKERN_BASE zlib iotracking ]
