diff --git a/cmake/kext.cmake b/cmake/kext.cmake
index 5f155bb7..fbb9873c 100644
--- a/cmake/kext.cmake
+++ b/cmake/kext.cmake
@@ -13,7 +13,7 @@ function(add_kext_bundle name)
 
     target_compile_definitions(${name} PRIVATE TARGET_OS_OSX KERNEL)
     target_compile_options(${name} PRIVATE $<$<COMPILE_LANGUAGE:CXX>:-fapple-kext>)
-    target_link_options(${name} PRIVATE "LINKER:-bundle")
+    target_link_options(${name} PRIVATE "LINKER:-kext")
     target_link_options(${name} PRIVATE "SHELL:-undefined dynamic_lookup")
 
     if(SL_KERNEL_PRIVATE)
