diff --git a/src/firehose/firehose_buffer.c b/src/firehose/firehose_buffer.c
index 45dbf04..24aa180 100644
--- a/src/firehose/firehose_buffer.c
+++ b/src/firehose/firehose_buffer.c
@@ -81,9 +81,9 @@
 #include <sys/types.h>
 #include <vm/vm_kern.h>
 #include <internal/atomic.h> // os/internal/atomic.h
-#include <firehose_types_private.h> // <firehose/firehose_types_private.h>
-#include <tracepoint_private.h> // <firehose/tracepoint_private.h>
-#include <chunk_private.h> // <firehose/chunk_private.h>
+#include <firehose/firehose_types_private.h>
+#include <firehose/tracepoint_private.h>
+#include <firehose/chunk_private.h>
 #include "os/firehose_buffer_private.h"
 #include "firehose_buffer_internal.h"
 #include "firehose_inline_internal.h"
