--- a/src/add-ons/kernel/bus_managers/pci/pci_info.cpp
+++ b/src/add-ons/kernel/bus_managers/pci/pci_info.cpp
@@ -12,7 +12,8 @@
 #include "pci.h"
 
 #define PCI_VERBOSE	1
-#if defined(__i386__) || defined(__x86_64__)
+#if (defined(__i386__) || defined(__x86_64__)) \
+	&& !defined(HAIKU_MOUNTIN_BUILD)
 // enabling it makes the pci bus_manager binary about 1MB
 // some other platforms have issues with floppy image size...
 // TODO: Move this define to BuildSetup?
